src/zarr/core/buffer/__init__.py

```diff
@@ -1,19 +1,10 @@
-from zarr.core.buffer.core import (
-    ArrayLike,
-    Buffer,
-    BufferPrototype,
-    NDArrayLike,
-    NDBuffer,
-    default_buffer_prototype,
-)
-from zarr.core.buffer.cpu import numpy_buffer_prototype
+import warnings
+
+from zarr.abc.buffer import Buffer, BufferPrototype, NDBuffer  # noqa: F401
 
-__all__ = [
-    "ArrayLike",
-    "Buffer",
-    "BufferPrototype",
-    "NDArrayLike",
-    "NDBuffer",
-    "default_buffer_prototype",
-    "numpy_buffer_prototype",
-]
+warnings.warn(
+    "The zarr.core.buffer module is deprecated and will be removed in a future version. "
+    "Please use zarr.buffer and zarr.abc.buffer modules instead.",
+    FutureWarning,
+    stacklevel=2,
+)
```

Comment:
put this in to prompt a discussion -- numcodecs depends on this module already... which sucks.

@normanrz - do you have thoughts on how to proceed? Options are:

1. keep `zarr.core` but document it as private. We've managed to do this in the Xarray project with few issues.
2. change the [import in `numcodecs`](https://github.com/zarr-developers/numcodecs/blob/6c0ea0f43788f22e4ad5d6655862c6c212a0dece/numcodecs/zarr3.py#L51-L54) and make a new release

Reply:
We can easily skip documenting some 'public' (without a `_`) modules, which would probably help: https://github.com/zarr-developers/zarr-python/blob/f9c20243d207835d15d7091e05e8ec82a265b7d1/docs/conf.py#L64-L76

Reply:
I would also lean towards 1. Changing numcodecs would almost certainly also require some backwards-compat shim in zarr-python.